# Chapter 2: An Array of Sequences

有些操作，对String, List, tables 是相同的（例如：`for element in list_a`）。这三种数据结构又称作Trains。

## 1. Overview of  Build-In Sequences

The standard library offers a rich selection of sequence types implemented in C:
- Container sequences: hold references to the objects they contain (可以理解成一个指针或是引用)
    - list, tuple, and collections.deque can hold items of different types.
- Flat sequences: store the value of each item within its own memory (直接存值)
    - primitive values: str, bytes, bytearray, memoryview, and array.array hold items of one type.
    
另外，可以根据一个sequence 是否是mutable，分为两类：
- mutable sequences
    - list, bytearray, array.array, collections.deque, and memoryview
- Immutable sequences
    - tuple, str, and bytes
   
<img src="figures/mutable_sequence.png" width="500"/>

   
最常用的sequence 是list，下面我们先从list 开始。

### 1.1 List Comprehension

In [1]:
list_a = [1,2,3,4,5]
list_b = []

#### 方法1: for loop

In [2]:
for element in list_a:
    list_b.append(str(element))
list_b

['1', '2', '3', '4', '5']

#### 方法2: List Comprehension

list comprehension 只做一件事：生成一个新的list

In [3]:
list_c = [str(element) for element in list_a]
list_c

['1', '2', '3', '4', '5']

#### Listcomps v.s. map and filter

Map and filter: more details in Chapter 5.

In [4]:
list_d = [str(element) for element in list_a if element > 2]
list_d

['3', '4', '5']

In [5]:
list_5 = list(filter(lambda c: int(c) > 2, map(str, list_a)))
list_5

['3', '4', '5']

#### 使用list comprehension 实现cartesian products

generate lists from the Cartesian product of two or more iterables.

In [11]:
colors = ['black', 'white']
sizes = ['S', 'M', 'L']
tshirts = [(color, size) for color in colors for size in sizes] # list of tuples
tshirts

[('black', 'S'),
 ('black', 'M'),
 ('black', 'L'),
 ('white', 'S'),
 ('white', 'M'),
 ('white', 'L')]

如果不用list comprehension，我们需要写两个for loop 嵌套

### 1.2 Generator Expression

用来生成其他类型的sequences (tuples, arrays)

Genexps use the same syntax as listcomps, but are enclosed in parentheses rather than brackets

下面我们看tuple

In [6]:
tuple_a = tuple(str(element) for element in list_a)
tuple_a

('1', '2', '3', '4', '5')

**The generator expression yields items one by one; a list with all six T-shirt variations is never produced in this example**.

如果只用一次，使用generator 会很快，尤其当item 很多的时候。

更多关于generator 的信息，参见Chapter 14.

In [7]:
sum([x*x for x in range(10)])

285

上面这种写法，内存里需要保存全量的数组。使用下面这种方法，Memory is conserved by using a generator expression instead:

In [8]:
sum((x*x for x in range(10)))

285

In [17]:
for tshirt in ('%s %s' % (c, s) for c in colors for s in sizes):
    print(tshirt)

black S
black M
black L
white S
white M
white L


有关generator 的更多介绍，参见Chapter 14.

## 2. Tuples Are Not Just Immutable Lists

#### mutable v.s. immutable
A mutable object can be changed after it's created, and an immutable object can't.

Tuple 的两个作用：
1. Tuple 可以看作是不可更改的list (immutable list)
2. 因为tuple 的顺序是不可以改的，所以tuple 可以看作"record" with no field name.

### 2.1 Tuple as immutable list

list 的所有方法，tuple 都支持 (除了增加和删除)，参见下表。

<img src="figures/tuple_as_immutable_list_1.png" width="500"/>
<img src="figures/tuple_as_immutable_list_2.png" width="500"/>

#### extend

In [11]:
list_1 = [1,2,3]
list_2 = [4,5,6]
list_1.__add__(list_2)

[1, 2, 3, 4, 5, 6]

In [12]:
tuple_1 = (1, 2, 3)
tuple_2 = (4, 5, 6)
tuple_1.__add__(tuple_2)

(1, 2, 3, 4, 5, 6)

#### slicing

In [13]:
list_1[:2]

[1, 2]

In [14]:
tuple_1[:2]

(1, 2)

### 2.2 Tuple as record

**tuple unpacking**

#### 1. parallel assignnment
- swapping
- return multiple value of a function

In [15]:
la_coordinates = (33.9425, -118.408056)

In [16]:
latitude = la_coordinates[0]
latitude

33.9425

In [17]:
longitude = la_coordinates[1]
longitude

-118.408056

以上是分别赋值，下面我们看parallel assignment

In [18]:
latitude, longitude = la_coordinates
print(latitude)
print(longitude)

33.9425
-118.408056


**第一个用处**：

parallel assignment 一个常用的场景是swapping without temporary variable

In [19]:
a = 1
b = 5
print ('a: ', a)
print('b: ', b)

# swap
c = a
a = b
b = c
print ('a: ', a)
print('b: ', b)

a:  1
b:  5
a:  5
b:  1


In [20]:
a = 1
b = 5
print ('a: ', a)
print('b: ', b)

# 使用parallel assignment 
b, a = a, b # 一行代码实现
print ('a: ', a)
print('b: ', b)

a:  1
b:  5
a:  5
b:  1


**第二个用处**：

enable functions to return multiple values in a way that is convenient to the caller.

In [21]:
def get_first_two_elements(list_a):
    return list_a[0], list_a[1]

In [22]:
list_a = [1, 2]

In [23]:
a, b = get_first_two_elements(list_a)
print(a)
print(b)

1
2


可以用一个**dummy variable** `_` 作为一个place holder，表明不在意某个变量

In [24]:
_, b = get_first_two_elements(list_a)
print(b)

2


In [25]:
import os
_, filename = os.path.split('/home/myname/paper.tex')

In [26]:
filename

'paper.tex'

可以使用 `*args` 来匹配多个items

In [27]:
a, b, *c = range(5)
print(a)
print(b)
print(c)

0
1
[2, 3, 4]


In [28]:
a, b, *c = range(3)
print(a)
print(b)
print(c)

0
1
[2]


In [29]:
a, b, *c = range(2)
print(a)
print(b)
print(c)

0
1
[]


`*args` 可以出现在任意位置

In [30]:
a, *b, c = range(5)
print(a)
print(b)
print(c)

0
[1, 2, 3]
4


`*args` 只能出现一次，否则会报错

In [31]:
a, *b, *c = range(2)
print(a)
print(b)
print(c)

SyntaxError: two starred expressions in assignment (<ipython-input-31-3678e28f7358>, line 1)

#### 2. Nested Tuple Unpacking

In [32]:
cities = [('Tokyo', 'JP', 36.933, (35.689722, 139.691667)),
          ('Delhi', 'IND', 21.935, (28.613889, 77.208889)),
          ('Mexico City', 'MX', 20.142, (19.43333, -99.133333))]

In [33]:
print('{} | {}'.format('', 'Longitude'))
for city, country, population, (latitude, longitude) in cities:
    # 只显示东半球的城市
    if longitude > 0:
        print ('{} | {}'.format(city, longitude))

 | Longitude
Tokyo | 139.691667
Delhi | 77.208889


#### 表格对齐

In [34]:
print('{:15} | {:^6}'.format('', 'Longitude'))
for city, country, population, (latitude, longitude) in cities:
    # 只显示东半球的城市
    if longitude > 0:
        print ('{:15} | {:9.4f}'.format(city, longitude))

                | Longitude
Tokyo           |  139.6917
Delhi           |   77.2089


Tuple 用来表示一个record 有一个问题，就是只能按顺序表示每个field 的意思，但是没有field name。
如果想加入field name，则需要一namedtuple 对象。

#### 3. Name Tuples

**注意** Named Tuple 对象所占内存大小与tuple 一样，因为field name 信息存在class 定义中而非object 的attribute 中。

In [35]:
from collections import namedtuple

In [36]:
# 定义一个named tuple 类
# 参数1: class name (city)
# 参数2: list of field name (iterable of strings / single space delimited string)
City = namedtuple('City', 'name country population coordinates') # Using single space delimited string
tokyo = City('Tokyo', 'JP', 36.933, (35.689722, 139.691667))

In [37]:
tokyo

City(name='Tokyo', country='JP', population=36.933, coordinates=(35.689722, 139.691667))

In [38]:
tokyo.coordinates # 使用.field_name 来访问每个field

(35.689722, 139.691667)

In [39]:
City2 = namedtuple('City', ['name', 'Country', 'Population', 'Coordinates']) # using iterable of strings
tokyo2 = City2('Tokyo', 'JP', 36.933, (35.689722, 139.691667))

In [40]:
tokyo2

City(name='Tokyo', Country='JP', Population=36.933, Coordinates=(35.689722, 139.691667))

namedtuple 有三个比较关键的attribtues 和methods:
- `_fields`: 获取所有field 信息 
- `_make(interable)`: 通过一个iterable 来构建一个namedtuple
- `_asDict()`: 将namedtuple 转换成一个`collection.OrderedDict` 对象 

In [41]:
tokyo._fields

('name', 'country', 'population', 'coordinates')

In [42]:
City._fields

('name', 'country', 'population', 'coordinates')

In [43]:
delhi_data = ('Delhi', 'IND', 21.935, (28.613889, 77.208889))

In [44]:
delhi = City._make(delhi_data)

In [45]:
delhi

City(name='Delhi', country='IND', population=21.935, coordinates=(28.613889, 77.208889))

In [46]:
delhi._asdict()

OrderedDict([('name', 'Delhi'),
             ('country', 'IND'),
             ('population', 21.935),
             ('coordinates', (28.613889, 77.208889))])

可以通过`asdict()`做显示上的优化

In [47]:
for key, value in delhi._asdict().items():
    print(key + ':', value)

name: Delhi
country: IND
population: 21.935
coordinates: (28.613889, 77.208889)


## 3. Slicing and Concatenation

- slicing: 对一个sequence 取一部分(做减法)
- concatenation: 将多个sequence 合并(做加法)

### 3.1 Slicing
sequence types (e.g., list, tuple, str) 都支持slicing 操作，这一节我们介绍一下slicing 的高级用法。

#### 1. 基本操作

`[start:end:step]`

In [48]:
l = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

In [49]:
l[:2]

[1, 2]

In [50]:
l[2:]

[3, 4, 5, 6, 7, 8, 9, 10]

#### step

In [51]:
s = 'hello world!'

In [52]:
s[::3] # 不限头尾，step = 3

'hlwl'

In [53]:
s[::-3] # 不限头尾，从右向左 step = 3

'!r l'

In [54]:
s[2::2] # 不限头尾，从右向左 step = 3

'lowrd'

In [55]:
s[2::-2] # 不限头尾，从右向左 step = 3

'lh'

#### 2. slice objects

slice 其实实现是创建了一个slice 对象`slice(start, stop, step)`.
创建slice 对象在许多场景中很handy.


In [56]:
invoice = """
... 0.....6.................................40........52...55........
... 1909  Pimoroni PiBrella                 $17.50      3  $52.50
... 1489  6mm Tactile Switc x20             $4.95       2  $9.90
... 1510  Panavise Jr. - PV-201             $28.00      1  $28.00
... 1601  PiTFT Mini Kit 320x240            $34.95      1  $34.95
... """

In [57]:
invoice

'\n0.....6.................................40........52...55........\n1909  Pimoroni PiBrella                 $17.50      3  $52.50\n1489  6mm Tactile Switc x20             $4.95       2  $9.90\n1510  Panavise Jr. - PV-201             $28.00      1  $28.00\n1601  PiTFT Mini Kit 320x240            $34.95      1  $34.95\n'

In [58]:
SKU = slice(0, 6)
DESCRIPTION = slice(6, 40)
UNIT_PRICE = slice(40, 52)
QUANTITY = slice(52, 55)
ITEM_TOTAL = slice(55, None)

In [59]:
line_items = invoice.split('\n')[2:] # 从第三行开始 去除前两行文件头

In [60]:
line_items

['1909  Pimoroni PiBrella                 $17.50      3  $52.50',
 '1489  6mm Tactile Switc x20             $4.95       2  $9.90',
 '1510  Panavise Jr. - PV-201             $28.00      1  $28.00',
 '1601  PiTFT Mini Kit 320x240            $34.95      1  $34.95',
 '']

In [61]:
for item in line_items:
    print(item[UNIT_PRICE], item[DESCRIPTION])

$17.50       Pimoroni PiBrella                 
$4.95        6mm Tactile Switc x20             
$28.00       Panavise Jr. - PV-201             
$34.95       PiTFT Mini Kit 320x240            
 


#### 3. multi-dimensional slicing



In [62]:
import numpy as np
data = np.array([[1,2,3,4,5],
                 [6,7,8,9,10],
                 [11,12,13,14,15]])

In [63]:
data[:2, :2]

array([[1, 2],
       [6, 7]])

#### ellipsis `...`

在高维数组中(例如: tensor)，用ellipsis 表示省略。

`x[i, ...]` is a shortcut for `x[i, :, :, :,]`

In [64]:
x = np.array([[[111,112,113,114,115], [121,122,123,124,125]], [[211,212,213,214,215], [221,222,223,224,225]]])


In [65]:
x.ndim

3

In [66]:
x.shape

(2, 2, 5)

In [67]:
x[0, :, :,]

array([[111, 112, 113, 114, 115],
       [121, 122, 123, 124, 125]])

In [68]:
x[0, ...]

array([[111, 112, 113, 114, 115],
       [121, 122, 123, 124, 125]])

#### 4. Assigning to slice

注意，等号的右边必须是一个interable object, 例如list.

In [69]:
l = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

In [70]:
l[0:5] = [1] # 重新赋值前5个值

In [71]:
l

[1, 6, 7, 8, 9, 10]

In [72]:
l[0:5] = 1 # error

TypeError: can only assign an iterable

In [73]:
del l[1:] # 删除除了第一个元素的所有元素

In [74]:
l

[1]

### 3.2 Concatenation

#### 1. 使用`+`

必须是相同类型(sequence type)，例如，不能把一个String 和一个list 相加。

In [75]:
l1 = [1,2,3]
l2 = [4,5,6]

In [76]:
l1 + l2

[1, 2, 3, 4, 5, 6]

In [77]:
l1 + 'hello' # 不同类型不能相加

TypeError: can only concatenate list (not "str") to list

In [79]:
l1 + ['hello'] # 不要求元素是相同类型

[1, 2, 3, 'hello']

#### 2. 使用`*`

注意，both `+` and `*` 都会创建一个新的对象，不会改变原始对象

In [80]:
l3 = l1 * 3
l3

[1, 2, 3, 1, 2, 3, 1, 2, 3]

In [81]:
'hello ' * 3

'hello hello hello '

In [82]:
l3[0] = 0

In [83]:
l3 # 修改l3

[0, 2, 3, 1, 2, 3, 1, 2, 3]

In [84]:
l1 # l1 不会有变化

[1, 2, 3]

In [85]:
my_list = [[]]
my_list

[[]]

In [86]:
my_list_of_list = my_list * 3
my_list_of_list

[[], [], []]

In [87]:
my_list_of_list[0] = [1,2,3]

In [88]:
my_list_of_list

[[1, 2, 3], [], []]

In [89]:
my_list

[[]]

#### 3. Building Lists of Lists

In [90]:
# 通过list comprehension 创建一个空的棋盘
board = [['_'] * 3 for i in range(3)]

In [91]:
board

[['_', '_', '_'], ['_', '_', '_'], ['_', '_', '_']]

In [92]:
board[0][1] = 'x'

In [93]:
board

[['_', 'x', '_'], ['_', '_', '_'], ['_', '_', '_']]

上面的做法等同于以下方法:

In [94]:
board4=[]
for i in range(3):
    row = ['_'] * 3
    board4.append(row)

In [95]:
board4

[['_', '_', '_'], ['_', '_', '_'], ['_', '_', '_']]

In [96]:
board4[0][1] = 'x'

In [97]:
board4

[['_', 'x', '_'], ['_', '_', '_'], ['_', '_', '_']]

#### pitfall: 下面演示一个错误的做法

In [98]:
board2 = [['_'] * 3] * 3 

In [99]:
board2

[['_', '_', '_'], ['_', '_', '_'], ['_', '_', '_']]

In [100]:
board2[0][1] = 'x'

In [101]:
board2

[['_', 'x', '_'], ['_', 'x', '_'], ['_', 'x', '_']]

the outer list is made of three references to the same inner list. So all rows are aliases to refer to the same object.

上面的代码，其实等同于下面代码：

In [103]:
row = ['_'] * 3
board3 = []
for i in range(3):
    board3.append(row) # 相同的对象row 被append 了三次

In [104]:
board3

[['_', '_', '_'], ['_', '_', '_'], ['_', '_', '_']]

In [105]:
board3[0][1] = 'x'

In [106]:
board3

[['_', 'x', '_'], ['_', 'x', '_'], ['_', 'x', '_']]

我们在第八章重点介绍references and mutable objects.

介绍了`+` 和`*`, 我们下面介绍`+=` 和`*=` operators

## 4. Augmented Assignment with Sequences

### 4.1 Augmented Addition `+=`

`+=` 背后的实现是`__iadd__`方法(in-place addition 原地修改). 如果`__iadd__` 没有实现，python 会再去调用`__add__` 方法。

#### 例如：`a += b`
- 如果a 实现了`__iadd__` 方法，则a 的值直接被修改成a+b 的值, 而a 的引用不变
- 如果a 没有实现`__iadd__` 方法，则首先计算a+b，生成一个新的对象，然后再把这个对象的引用赋值给a

一般mutable sequence 最好都实现`__iadd__` 方法。

In [107]:
l = [1, 2, 3]

In [108]:
id(l)

4515613768

In [109]:
l += [2]

In [110]:
l

[1, 2, 3, 2]

In [111]:
id(l) # in-place 对象不变

4515613768

In [112]:
l = l + [2] # repeated concatenation: inefficient

In [113]:
l

[1, 2, 3, 2, 2]

In [114]:
id(l) # 生成了一个新的对象

4515610760

#### Puzzler: 有趣的代码片段

假设我们执行以下代码片段：

```python
a = (1, 2, [3, 4])
a[2] += [5, 6]
```

因为a 是一个tuple，tuple是immutable 的（即不能修改）。
现在对`a[2]` 做`+=` 操作，理论上，这个操作是in-place 的，也就是没有更改`a[2]`的引用，但是要修改的值。

所以执行这个代码，到底能不能修改成功？

In [115]:
a = (1, 2, [3, 4])
a[2] += [5, 6]

TypeError: 'tuple' object does not support item assignment

In [116]:
a

(1, 2, [3, 4, 5, 6])

首先，执行过程中报错了，果然，tuple 是immutable 不能修改。

但是，我们发现，内容其实已经修改成功了。

通过pythontutor.com 我们可以看到其执行步骤。

<img src="figures/tuple_immutable1.png" width="800"/>
<img src="figures/tuple_immutable2.png" width="800"/>

上面的功能，主要是赋值的动作是非法的，但其实我们是in-place，并没有真正赋值。我们可以通过extend 来修改其内容。

In [225]:
a[2].extend([7, 8])

In [226]:
a

(1, 2, [3, 4, 5, 6, 7, 8])

从上面的例子，我们可以看出：
- 把mutable 对象放到tuple 中并不是一个好的想法
- augmented assignment 并不是一个atomic operation（报错了，还修改成功了）

### 4.2 `*=`

同`+=`, 实现了`__imul__`方法。

## 5. Sorting

### 5.1 比较两种sorting 方法

Python 有两种sort 方法：第一对每个list 调用`lst.sort()`方法，第二使用`sorted(lst)` 方法。

- `list.sort` sort list in place --- returns None (没有创建新的list).

- 相反，build-in 函数 `sorted` create new list and return it. 可以接受任何一个iterable 对象作为输入，输出都是一个list.

这两个方法背后的排序算法是timsort - an adaptive algorithm that switches from insertion sort to merge sort strategies, depending on how ordered the data is.

#### basic sorting

In [117]:
fruits = ['grape', 'apple', 'orange']  # 创建一个list

In [118]:
fruits

['grape', 'apple', 'orange']

In [119]:
fruits.sort()  # 方法1

In [120]:
fruits

['apple', 'grape', 'orange']

In [121]:
result = list.sort(fruits) # in-place sort
print(result)

None


In [122]:
result = sorted(fruits, reverse=True)  # 方法2

In [123]:
fruits # not changes

['apple', 'grape', 'orange']

In [124]:
result # reverse order

['orange', 'grape', 'apple']

#### key Function

两种排序方法都支持key parameter: specify a function to be called on each list element prior to making comparisons.

In [238]:
result.sort(key = len)


In [240]:
result # 先按长度排，再按字符倒排

['grape', 'apple', 'orange']

In [241]:
result2 = sorted(result, key=len, reverse=True)

In [242]:
result2

['orange', 'grape', 'apple']

The value of the key parameter should be a function that takes a single argument and returns a key to use for sorting purposes.

In [243]:
student_tuples = [
    ('john', 'A', 15),
    ('jane', 'B', 12),
    ('dave', 'B', 10)]
sorted(student_tuples, key=lambda student: student[2])   # sort by age


[('dave', 'B', 10), ('jane', 'B', 12), ('john', 'A', 15)]

In [245]:
sorted(student_tuples, key=lambda student: student[1])   # sort by score

[('john', 'A', 15), ('jane', 'B', 12), ('dave', 'B', 10)]

#### Operator Module Function

In [246]:
from operator import itemgetter

sorted(student_tuples, key=itemgetter(2))

[('dave', 'B', 10), ('jane', 'B', 12), ('john', 'A', 15)]

itemgetter 可以获取tuple 的元素。
同样，可以使用attrgetter 来获取一个对象的attribute

```python
from operator import itemgetter, attrgetter

student_objects = [
    Student('john', 'A', 15),
    Student('jane', 'B', 12),
    Student('dave', 'B', 10)]

sorted(student_objects, key=attrgetter('age'))
```

使用operator module functions 可以实现多级排序。

In [247]:
sorted(student_tuples, key=itemgetter(1,2))  # 首先对成绩排序，然后对age 排序

[('john', 'A', 15), ('dave', 'B', 10), ('jane', 'B', 12)]

#### Ascending and Descending

reverse parameter with a boolean value (参见上面的例子)。

#### sort stability

stable means when multiple records have the same key, their original order is preserved.

In [248]:
data = [('red', 1), ('blue', 1), ('red', 2), ('blue', 2)]
sorted(data, key=itemgetter(0))

[('blue', 1), ('blue', 2), ('red', 1), ('red', 2)]

#### complex sort

e.g., to sort the student data by descending grade and then ascending age

注意，在这个例子里，从String 来看，B>A, 但是实际含义A>B.

In [250]:
s = sorted(student_tuples, key=itemgetter(2))

In [252]:
sorted(s, key=itemgetter(1), reverse=True)

[('dave', 'B', 10), ('jane', 'B', 12), ('john', 'A', 15)]

我们可以把上面的功能用一个函数实现

In [254]:
def complex_sort (xs, specs):
    for key, reverse in reversed(specs):
        xs.sort(key=itemgetter(key), reverse=reverse)
    
    return xs

In [258]:
specs = ((1, True), (2, False))  # 根据tuple 的第二个元素reverse 排序， 根据tuple 的第三个元素正序排序。

In [257]:
complex_sort(student_tuples, specs)

[('dave', 'B', 10), ('jane', 'B', 12), ('john', 'A', 15)]

More information: https://docs.python.org/3/howto/sorting.html

### 5.2 使用bisect 管理ordered sequence

- `bisect` - quick find from a sorted list
- `insort` - insert to a sorted list

这两个方法使用二分查找方法。

#### 5.2.1 search with bisect

`bisect(haystack, needle)` - binary search: haystack must be ordered.

这个方法的返回值可以是作为index argument to `haystack.insert(index, needle)`.

In [127]:
import bisect
import sys

In [128]:
HAYSTACK = [1, 4, 5, 6, 8, 12, 15, 20, 21, 23, 23, 26, 29, 30]
NEEDLES = [0, 1, 2, 5, 8, 10, 22, 23, 29, 30, 31]

In [136]:
"""
返回值：index - where needle can be insert while keep the haystack ordered
all items appearing up to that position are less than or equal to needle
"""
bisect.bisect(HAYSTACK, 100)  # 列表长度14， index 0 ~ 13

14

In [138]:
bisect.bisect(HAYSTACK, 0)

0

这个返回值可以看作insert 的index,

```Python
index = bisect(haystack, needle)
haystack.insert(index, needle)
```

上述两行代码等同于：

```Python
haystack.insort(haystack, needle)
```

In [ ]:
# HAYSTACK.insert(0, 0)

In [129]:
ROW_FMT = '{0:2d} @ {1:2d} {2}{0:<2d}' # 输出的format

In [147]:
def demo(bisect_fn):
    for needle in reversed(NEEDLES):
        position = bisect_fn(HAYSTACK, needle)
        offset = ' ' * 3 + position * '  |'
        print(ROW_FMT.format(needle, position, offset))

In [148]:
bisect_fn = bisect.bisect_left

print('DEMO:', bisect_fn.__name__)
print('haystack ->', ' '.join('%2d' % n for n in HAYSTACK))
demo(bisect_fn)

DEMO: bisect_left
haystack ->  1  4  5  6  8 12 15 20 21 23 23 26 29 30
31 @ 14      |  |  |  |  |  |  |  |  |  |  |  |  |  |31
30 @ 13      |  |  |  |  |  |  |  |  |  |  |  |  |30
29 @ 12      |  |  |  |  |  |  |  |  |  |  |  |29
23 @  9      |  |  |  |  |  |  |  |  |23
22 @  9      |  |  |  |  |  |  |  |  |22
10 @  5      |  |  |  |  |10
 8 @  4      |  |  |  |8 
 5 @  2      |  |5 
 2 @  1      |2 
 1 @  0    1 
 0 @  0    0 


`bisect` is actually an alias for `bisect_right`.

`bisect_left` and `bisect_right` 之间的区别在于，当needle 等于某个元素的时候：
- `bisect_left` 返回这个元素的index
- `bisect_right` 返回这个元素的index 之后的可插入点（对于没有重复元素的场景，为当前元素的index+1） 


In [149]:
bisect_fn = bisect.bisect

print('DEMO:', bisect_fn.__name__)
print('haystack ->', ' '.join('%2d' % n for n in HAYSTACK))
demo(bisect_fn)

DEMO: bisect
haystack ->  1  4  5  6  8 12 15 20 21 23 23 26 29 30
31 @ 14      |  |  |  |  |  |  |  |  |  |  |  |  |  |31
30 @ 14      |  |  |  |  |  |  |  |  |  |  |  |  |  |30
29 @ 13      |  |  |  |  |  |  |  |  |  |  |  |  |29
23 @ 11      |  |  |  |  |  |  |  |  |  |  |23
22 @  9      |  |  |  |  |  |  |  |  |22
10 @  5      |  |  |  |  |10
 8 @  5      |  |  |  |  |8 
 5 @  3      |  |  |5 
 2 @  1      |2 
 1 @  1      |1 
 0 @  0    0 


对于插入的结果，因为相同，插入在当前位置，还是插入在下一个位置，make no difference。但是，对于不同数据类型比较，就不一样了。

In [156]:
sample_haystack = [0, 1, 2]
index = bisect.bisect(sample_haystack, 1.0)
print('index: ', index)
sample_haystack.insert(index, 1.0)
print(sample_haystack)

index:  2
[0, 1, 1.0, 2]


In [157]:
sample_haystack = [0, 1, 2]
index = bisect.bisect_left(sample_haystack, 1.0)
print('index: ', index)
sample_haystack.insert(index, 1.0)
print(sample_haystack)

index:  1
[0, 1.0, 1, 2]


bisect 的一个应用：to convert test scores to letter grades

ref: https://docs.python.org/3/library/bisect.html

In [158]:
def grade(score, breakpoints=[60, 70, 80, 90], grades='FDCBA'):
    i = bisect.bisect(breakpoints, score)
    return grades[i]

In [159]:
[grade(score) for score in [33, 99, 77, 70, 89, 90, 100]]

['F', 'A', 'C', 'C', 'B', 'A', 'A']

#### 5.2.2 Inserting with `bisect.insort`

Sorting is expensive, so once you have a sorted sequence, it’s good to keep it that way.

`insort(seq, item)` inserts item into seq so as to keep seq in ascending order.

In [160]:
import bisect
import random

In [163]:
SIZE = 7
random.seed(42)
my_list = []
for i in range(SIZE):
    new_item = random.randrange(SIZE*2)
    bisect.insort(my_list, new_item)
    print('%2d ->' % new_item, my_list)

10 -> [10]
 1 -> [1, 10]
 0 -> [0, 1, 10]
11 -> [0, 1, 10, 11]
 4 -> [0, 1, 4, 10, 11]
 3 -> [0, 1, 3, 4, 10, 11]
 3 -> [0, 1, 3, 3, 4, 10, 11]


Like bisect, `insort` takes optional `lo`, `hi` arguments to limit the search to a subsequence.

## 6. When list is not the answer

当list 的元素都是number 的时候，用array 是一个更好的选择。

### 6.1 Arrays

If the list will only contain numbers, an array.array is more efficient than a list.

In [172]:
from array import array
from random import random
import time

In [165]:
floats = array('d', (random() for i in range(10**7)))  # 生成10,000,000个随机float

上面的d是一个typecode，代表double 数据类型。typecode 参见下表。

| Type code |       C Type       |    Python Type    |   |
|:---------:|:------------------:|:-----------------:|:-:|
|     b     |     signed char    |        int        | 1 |
|     B     |    unsigned char   |        int        | 1 |
|     u     |       UNICODE      | UNICODE Character | 2 |
|     h     |    sighed short    |        int        | 2 |
|     H     |   unsigned short   |        int        | 2 |
|     i     |     signed int     |        int        | 2 |
|     I     |    unsigned int    |        int        | 2 |
|     l     |     singed long    |        int        | 4 |
|     L     |    unsinged long   |        int        | 4 |
|     q     |  signed long long  |        int        | 8 |
|     Q     | unsigned long long |        int        | 8 |
|     f     |        float       |       float       | 4 |
|     d     |       double       |       float       | 8 |

In [166]:
floats[-1]

0.1703067320049848

In [176]:
# 将array写入bin 文件的时间
start_time = time.time()
fp = open('floats.bin', 'wb')
floats.tofile(fp)
fp.close()
end_time = time.time()
print('Write bin file time: ',end_time - start_time)

Write bin file time:  0.2761878967285156


In [177]:
# 将bin 文件读入内存，构建array的时间

floats2 = array('d')

start_time2 = time.time()
fp = open('floats.bin', 'rb')
floats2.fromfile(fp, 10**7)
fp.close()
end_time2 = time.time()
print('Read bin file time: ', end_time2 - start_time2)


Read bin file time:  0.19189882278442383


In [184]:
type(floats2)

array.array

In [187]:
# 查看bin文件大小

import os
bin_size = os.path.getsize("floats.bin")
bin_size

80000000

In [186]:
floats2[-1]  # 读出的和写入的值相同

0.1703067320049848

In [181]:
t_start = time.time()
with open('floats.txt', 'w') as f:
    for num in floats:
        f.write(str(num) + '\n')
t_end = time.time()

t_end - t_start


14.123858213424683

In [182]:
t_start = time.time()
floats3 = array('d')
with open('floats.txt', 'r') as f:
    floats3 = f.read().splitlines()
t_end = time.time()

t_end - t_start

2.325744867324829

In [183]:
type(floats3)

list

In [185]:
bin_size = os.path.getsize("floats.txt")
bin_size

192701341

可以看出，使用array 存成bin文件，大小为：80,000,000.因为一个floats 是8个字节，一共10,000,000个floats，总共80,000,000 字节。
如果存成txt，则需要192,701,341，大致是bin 的2.5倍。

同样，txt 文件的读取时间是bin 文件读取时间的75倍。txt 文件的写入时间是bin 文件写入时间的50倍。所以对数值类型的list，使用array 更有效。

之前我们说过，array 可以看作数值列表，下面我们就比较一下array 和list.

<img src="figures/array_and_list.png" width="500"/>



### 6.2 Memory Views

The built-in memorview class is a shared-memory sequence type that lets you handle
slices of arrays without copying bytes.

ref: https://eli.thegreenplace.net/2011/11/28/less-copies-in-python-with-the-buffer-protocol-and-memoryviews/

### 6.3 Numpy and Scipy

NumPy and SciPy are the reason why Python
became mainstream in scientific computing applications.

In [188]:
import numpy
a = numpy.arange(12)
a

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11])

In [189]:
type(a)

numpy.ndarray

In [190]:
a.shape

(12,)

In [191]:
a.shape = 3, 4  # 类似于reshape

In [192]:
a.shape

(3, 4)

In [193]:
a[2]

array([ 8,  9, 10, 11])

In [195]:
a[2, 1]

9

In [197]:
a[:, 1]  # 第二列

array([1, 5, 9])

In [198]:
a.transpose()  # transpose 

array([[ 0,  4,  8],
       [ 1,  5,  9],
       [ 2,  6, 10],
       [ 3,  7, 11]])

#### NumPy also supports high-level operations for loading, saving, and operating on all elements of a numpy.ndarray: 

In [210]:
from time import perf_counter as pc

t0=pc()

floats = numpy.loadtxt('floats.txt')

pc() - t0

39.48942466801964

可以看出，读取txt 文件需要40s

In [211]:
floats.shape  # 检查数据完整性

(10000000,)

In [212]:
floats[:10]

array([0.13953793, 0.10249518, 0.74066774, 0.54536653, 0.59049251,
       0.03178268, 0.09369524, 0.23266089, 0.60201873, 0.56124506])

In [213]:
floats *= 0.5

In [214]:
floats[:10]

array([0.06976896, 0.05124759, 0.37033387, 0.27268327, 0.29524626,
       0.01589134, 0.04684762, 0.11633045, 0.30100936, 0.28062253])

In [215]:
t0 = pc()
numpy.save('floats-numpy', floats)  # save to a .npy file
pc() - t0

0.47182577988132834

In [216]:
t0 = pc()

floats2 = numpy.load('floats-numpy.npy', 'r+')
pc() - t0

0.013532911892980337

我们可以看出，读一个npy 大概需要0.01秒左右，比之前读取txt 要快很多

In [217]:
floats2 *= 2

In [218]:
floats2[:10]

memmap([0.13953793, 0.10249518, 0.74066774, 0.54536653, 0.59049251,
        0.03178268, 0.09369524, 0.23266089, 0.60201873, 0.56124506])

## 7. Queues

之前我们介绍了plain old list， 下面我们来看看队列

In [220]:
lst = [1, 2, 3]


[1, 2, 3, 4]

#### LIFO (Last In First Out)

In [223]:
lst.append(4)  # 假如一个元素
lst

[1, 2, 3, 4]

In [224]:
lst.pop()  # 弹出最后一个元素

4

In [225]:
lst  # 查看结果

[1, 2, 3]

#### FIFO (First In First Out)

In [226]:
lst.pop(0)  # 弹出最后一个元素

1

In [227]:
lst

[2, 3]

问题是，每次弹出（或插入）第一个元素效率很低。

The class collections.deque is a thread-safe double-ended queue designed for fast
inserting and removing from both ends.

a deque can be bounded - created
with a maximum length—and then, when it is full, it discards items from the opposite
end when you append new ones

In [228]:
from collections import deque

In [230]:
dq = deque(range(10), maxlen=10)
dq

deque([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [231]:
"""
Rotating with n > 0 takes items from the right end and prepends them to the left; 
when n < 0 items are taken from left and appended to the right.
"""
dq.rotate(3)
dq

deque([7, 8, 9, 0, 1, 2, 3, 4, 5, 6])

In [232]:
dq.rotate(-4)
dq


deque([1, 2, 3, 4, 5, 6, 7, 8, 9, 0])

In [233]:
dq.appendleft(-1)

In [234]:
dq  # discard right-most value: 0

deque([-1, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [235]:
dq.extend([11, 22, 33])  # discard the three left-most values: -1, 1, 2
dq

deque([3, 4, 5, 6, 7, 8, 9, 11, 22, 33])

In [236]:
dq.extendleft([10, 20, 30, 40])
dq

deque([40, 30, 20, 10, 3, 4, 5, 6, 7, 8])

appending each successive item of the
iter argument to the left of the deque, therefore the final position of the items
is reversed.

#### 有关deque 的一些特点

- Note that deque implements most of the list methods, and adds a few specific to its design, like popleft and rotate.
- removing items from the middle of a deque is not as fast. It is really optimized for appending and popping from the ends.
- The append and popleft operations are atomic, so deque is safe to use as a LIFO queue in multithreaded applications without the need for using locks.

## Summary

- sequences
    - mutable v.s. immutable
    - flat v.s. container
- list comprehension and generator expressions
- tuple plays two roles:
    - records with unnamed fields
        - tuple unpacking
        - * syntax
        - named tuples
    - immutable list
- Sequence slicing
    - ellipsis notation for multi-dimentional slicing
    - slice object
- Repeated concatenation with *
- Augmented assignment with += and *=
    - immutable sequence: return new sequence
    - mutable sequence: in-place
- sort() and sorted()
- bisect.bisect and bisect.insort
- array.array
- collections.deque

## Appendix

### A1: Why numbering should start at zero

ref: https://www.cs.utexas.edu/users/EWD/transcriptions/EWD08xx/EWD831.html

### A2: MemoryView Tutorial

ref: https://eli.thegreenplace.net/2011/11/28/less-copies-in-python-with-the-buffer-protocol-and-memoryviews/